apps/aehttp/src/aehttp_cache_etag.erl

```diff
@@ -0,0 +1,174 @@
+%% Implements generate/3 similar to cowboy_rest generate_etag/2 callback
+%% customized with additional OperationId (first) argument for easy switching
+
+-module(aehttp_cache_etag).
+
+-export([generate/3]).
+
+-type cowboy_etag_res() :: {undefined | {strong, binary()}, cowboy_req:req(), any()}.
+
+%%%===================================================================
+%%% External API
+
+-spec generate(atom(), cowboy_req:req(), any()) -> cowboy_etag_res().
+generate('GetTopBlock', Req, State) ->
+    case aec_chain:top_block() of
+        undefined ->
+            {undefined, Req, State};
+        Block ->
+            ETag = etag_block(Block),
+            {{strong, ETag}, Req, State}
+    end;
+
+generate('GetCurrentKeyBlock', Req, State) ->
+    case aec_chain:top_key_block() of
+        {ok, Block} ->
+            ETag = etag_block(Block),
+            {{strong, ETag}, Req, State};
+        _ ->
+            {undefined, Req, State}
+    end;
+
+generate('GetKeyBlockByHash', Req, State) ->
+    Hash = cowboy_req:binding(hash, Req),
+    {{strong, Hash}, Req, State};
+
+generate('GetKeyBlockByHeight', Req, State) ->
+    handle_block_height(Req, State);
+
+generate('GetMicroBlockHeaderByHash', Req, State) ->
+    Hash = cowboy_req:binding(hash, Req),
+    {{strong, Hash}, Req, State};
+
+generate('GetMicroBlockTransactionsByHash', Req, State) ->
+    Hash = cowboy_req:binding(hash, Req),
+    {{strong, Hash}, Req, State};
+
+generate('GetMicroBlockTransactionByHashAndIndex', Req, State) ->
+    Hash = cowboy_req:binding(hash, Req),
+    {{strong, Hash}, Req, State};
+
+generate('GetMicroBlockTransactionsCountByHash', Req, State) ->
+    Hash = cowboy_req:binding(hash, Req),
+    {{strong, Hash}, Req, State};
+
+generate('GetGenerationByHash', Req, State) ->
+    handle_generation_hash(Req, State);
+
+generate('GetGenerationByHeight', Req, State) ->
+    handle_generation_height(Req, State);
+
+generate('GetTransactionByHash', Req, State) ->
+    Hash = cowboy_req:binding(hash, Req),
+    {{strong, Hash}, Req, State};
+
+generate('GetTransactionInfoByHash', Req, State) ->
+    Hash = cowboy_req:binding(hash, Req),
```

Comment:
I see the `height` of the `ContractCallObject` is part of the hash so shall be ok.